csharp/src/Ice/ConnectionI.cs

```diff
@@ -1647,31 +1648,40 @@ private void setState(int state)
                 case StateActive:
                 {
                     //
-                    // Can only switch from holding or not validated to
-                    // active.
+                    // Can only switch to active from holding or not validated.
                     //
                     if (_state != StateHolding && _state != StateNotValidated)
                     {
                         return;
                     }
+
+                    if (_maxDispatches > 0 && _dispatchCount >= _maxDispatches)
+                    {
+                        // Don't resume reads if the max dispatch count is reached.
+                        return;
+                    }
+
                     _threadPool.register(this, SocketOperation.Read);
                     break;
                 }
 
                 case StateHolding:
                 {
                     //
-                    // Can only switch from active or not validated to
-                    // holding.
+                    // Can only switch to holding from active or not validated.
```

Comment:
This comment is confusing. If this is an impossible state transition, we should assert, not "if ... return".